File: scripts/seed_db.py
#!/usr/bin/env python3
"""
Database seed script for NoteHub.
Creates sample users, notes, and tasks for testing.
"""

import os
import sys

# Add the src directory to the path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

from werkzeug.security import generate_password_hash
from datetime import datetime, timedelta
import random


def seed_database():
    """Seed the database with sample data."""
    from src.notehub import create_app, db
    from src.notehub.models import User, Note, Tag, Task

    app = create_app()

    with app.app_context():
        print("🌱 Starting database seeding...")

        # Create tables if they don't exist
        db.create_all()

        # Check if admin user already exists
        admin = User.query.filter_by(username='admin').first()
        if admin:
            print("⚠️  Admin user already exists. Skipping user creation.")
        else:
            # Create admin user
            admin_password = os.environ.get('NOTES_ADMIN_PASSWORD', 'admin123456789')
            admin = User(
                username='admin',
                email='admin@notehub.local',
                password=generate_password_hash(admin_password)
            )
            db.session.add(admin)
            print(f"✅ Created admin user (password: {admin_password})")

        # Create demo user
        demo = User.query.filter_by(username='demo').first()
        if demo:
            print("⚠️  Demo user already exists.")
        else:
            demo = User(
                username='demo',
                email='demo@notehub.local',
                password=generate_password_hash('demo123456789')
            )
            db.session.add(demo)
            print("✅ Created demo user (password: demo123456789)")

        db.session.commit()

        # Get the admin user for notes/tasks
        admin = User.query.filter_by(username='admin').first()

        # Create sample tags
        tag_names = ['Work', 'Personal', 'Ideas', 'Important', 'Todo', 'Reference']
        tags = {}
        for tag_name in tag_names:
            existing_tag = Tag.query.filter_by(name=tag_name, user_id=admin.id).first()
            if existing_tag:
                tags[tag_name] = existing_tag
            else:
                tag = Tag(name=tag_name, user_id=admin.id)
                db.session.add(tag)
                tags[tag_name] = tag
        
        db.session.commit()
        print(f"✅ Created {len(tag_names)} tags")

        # Sample notes data
        sample_notes = [
            {
                'title': 'Welcome to NoteHub! 🎉',
                'content': '''# Welcome to NoteHub!

This is your personal note-taking application. Here are some features you can explore:

## Features
- 📝 **Rich Markdown Support** - Write notes with full Markdown formatting
- 🏷️ **Tags** - Organize your notes with custom tags
- ⭐ **Favorites** - Mark important notes as favorites
- 📌 **Pin Notes** - Keep important notes at the top
- 🗂️ **Archive** - Archive old notes without deleting them
- 🔍 **Search** - Quickly find notes by title or content
- 📱 **Mobile Friendly** - Works great on all devices

## Getting Started
1. Click "New Note" to create your first note
2. Use Markdown for formatting
3. Add tags to organize your notes
4. Star your favorite notes

Enjoy using NoteHub!''',
                'tags': ['Important', 'Reference'],
                'pinned': True,
                'favorite': True,
            },
            {
                'title': 'Meeting Notes - Project Kickoff',
                'content': '''# Project Kickoff Meeting

**Date:** Today
**Attendees:** Team Alpha

## Agenda
1. Project overview
2. Timeline discussion
3. Resource allocation
4. Next steps

## Key Decisions
- Sprint duration: 2 weeks
- Daily standups at 9 AM
- Code reviews required for all PRs

## Action Items
- [ ] Set up project repository
- [ ] Create initial backlog
- [ ] Schedule sprint planning''',
                'tags': ['Work', 'Important'],
                'pinned': False,
                'favorite': True,
            },
            {
                'title': 'Book Recommendations 📚',
                'content': '''# Books to Read

## Currently Reading
- "Clean Code" by Robert C. Martin

## Up Next
1. "The Pragmatic Programmer"
2. "Design Patterns"
3. "Refactoring"

## Completed
- ✅ "JavaScript: The Good Parts"
- ✅ "You Don't Know JS"''',
                'tags': ['Personal', 'Reference'],
                'pinned': False,
                'favorite': False,
            },
            {
                'title': 'API Design Notes',
                'content': '''# REST API Design Best Practices

## Naming Conventions
- Use nouns for resources: `/users`, `/notes`
- Use plural forms
- Use hyphens for multi-word paths

## HTTP Methods
| Method | Usage |
|--------|-------|
| GET | Retrieve resources |
| POST | Create resources |
| PUT | Update resources |
| DELETE | Remove resources |

## Status Codes
- 200: Success
- 201: Created
- 400: Bad Request
- 401: Unauthorized
- 404: Not Found
- 500: Server Error''',
                'tags': ['Work', 'Reference'],
                'pinned': False,
                'favorite': False,
            },
            {
                'title': 'Weekend Plans',
                'content': '''# Weekend Todo

## Saturday
- [ ] Morning run
- [ ] Grocery shopping
- [ ] Call mom

## Sunday
- [ ] Meal prep
- [ ] Read book
- [ ] Movie night

## Notes
Don't forget to water the plants!''',
                'tags': ['Personal', 'Todo'],
                'pinned': False,
                'favorite': False,
            },
            {
                'title': 'Code Snippets Collection',
                'content': '''# Useful Code Snippets

## Python - Read JSON File
```python
import json

with open('data.json') as f:
    data = json.load(f)
```

## JavaScript - Fetch API
```javascript
const response = await fetch('/api/data');
const data = await response.json();
```

## SQL - Common Queries
```sql
SELECT * FROM users WHERE active = true;
```''',
                'tags': ['Work', 'Reference'],
                'pinned': False,
                'favorite': True,
            },
        ]

        # Create notes
        notes_created = 0
        for note_data in sample_notes:
            existing_note = Note.query.filter_by(
                title=note_data['title'],
                user_id=admin.id
            ).first()
            
            if not existing_note:
                note = Note(
                    title=note_data['title'],
                    content=note_data['content'],
                    user_id=admin.id,
                    pinned=note_data['pinned'],
                    favorite=note_data['favorite'],
                    created_at=datetime.utcnow() - timedelta(days=random.randint(0, 30))
                )
                
                # Add tags
                for tag_name in note_data['tags']:
                    if tag_name in tags:
                        note.tags.append(tags[tag_name])
                
                db.session.add(note)
                notes_created += 1

        db.session.commit()
        print(f"✅ Created {notes_created} sample notes")

        # Sample tasks
        sample_tasks = [
            {'title': 'Review pull requests', 'description': 'Check pending PRs on GitHub', 'priority': 'high', 'completed': False},
            {'title': 'Update documentation', 'description': 'Add API docs for new endpoints', 'priority': 'medium', 'completed': False},
            {'title': 'Fix login bug', 'description': 'Users report intermittent login issues', 'priority': 'high', 'completed': True},
            {'title': 'Setup CI/CD', 'description': 'Configure GitHub Actions for automated testing', 'priority': 'medium', 'completed': True},
            {'title': 'Design review meeting', 'description': 'Discuss new UI mockups with team', 'priority': 'low', 'completed': False},
            {'title': 'Database optimization', 'description': 'Add indexes to improve query performance', 'priority': 'medium', 'completed': False},
            {'title': 'Security audit', 'description': 'Review authentication and authorization', 'priority': 'high', 'completed': False},
            {'title': 'Write unit tests', 'description': 'Increase code coverage to 80%', 'priority': 'medium', 'completed': False},
        ]

        # Create tasks
        tasks_created = 0
        for task_data in sample_tasks:
            existing_task = Task.query.filter_by(
                title=task_data['title'],
                user_id=admin.id
            ).first()
            
            if not existing_task:
                due_date = datetime.utcnow() + timedelta(days=random.randint(1, 14))
                task = Task(
                    title=task_data['title'],
                    description=task_data['description'],
                    priority=task_data['priority'],
                    completed=task_data['completed'],
                    due_date=due_date if not task_data['completed'] else None,
                    user_id=admin.id,
                    created_at=datetime.utcnow() - timedelta(days=random.randint(0, 7))
                )
                db.session.add(task)
                tasks_created += 1

        db.session.commit()
        print(f"✅ Created {tasks_created} sample tasks")

        print("\n🎉 Database seeding completed successfully!")
        print("\n📋 Login credentials:")
        print(f"   Admin: admin / {os.environ.get('NOTES_ADMIN_PASSWORD', 'admin123456789')}")
        print("   Demo:  demo / demo123456789")


if __name__ == '__main__':
    seed_database()

File: backend/scripts/seed_db.js
/**
 * Database Seed Script
 *
 * Creates default admin and demo users with sample data.
 * Run with: npm run seed
 */
import dotenv from 'dotenv';

dotenv.config();

import bcrypt from 'bcryptjs';
import db from '../src/config/database.js';

async function seed() {
  try {
    console.log('🌱 Starting database seed...');

    await db.connect();
    await db.initSchema();

    // Get admin password from environment
    const adminPassword = process.env.NOTES_ADMIN_PASSWORD;
    if (!adminPassword || adminPassword.length < 12) {
      console.error('❌ NOTES_ADMIN_PASSWORD must be set and at least 12 characters');
      process.exit(1);
    }

    // Hash passwords
    const adminHash = await bcrypt.hash(adminPassword, 12);
    const demoHash = await bcrypt.hash('Demo12345678!', 12);

    // Create admin user
    const existingAdmin = await db.queryOne(`SELECT id FROM users WHERE username = ?`, ['admin']);
    if (!existingAdmin) {
      await db.run(
        `INSERT INTO users (username, password_hash, email, bio, is_admin) VALUES (?, ?, ?, ?, ?)`,
        ['admin', adminHash, 'admin@notehub.local', 'NoteHub Administrator', 1],
      );
      console.log('✅ Created admin user');
    } else {
      // Update existing admin to have is_admin flag
      await db.run(`UPDATE users SET is_admin = 1 WHERE username = ?`, ['admin']);
      console.log('ℹ️  Admin user already exists - updated is_admin flag');
    }

    // Create demo user
    const existingDemo = await db.queryOne(`SELECT id FROM users WHERE username = ?`, ['demo']);
    let demoUserId;
    if (!existingDemo) {
      const result = await db.run(
        `INSERT INTO users (username, password_hash, email, bio) VALUES (?, ?, ?, ?)`,
        ['demo', demoHash, 'demo@notehub.local', 'Demo user account for testing'],
      );
      demoUserId = result.insertId;
      console.log('✅ Created demo user');
    } else {
      demoUserId = existingDemo.id;
      console.log('ℹ️  Demo user already exists');
    }

    // Create sample tags
    const tags = ['work', 'personal', 'ideas', 'important', 'todo'];
    for (const tagName of tags) {
      const existing = await db.queryOne(`SELECT id FROM tags WHERE name = ?`, [tagName]);
      if (!existing) {
        await db.run(`INSERT INTO tags (name) VALUES (?)`, [tagName]);
      }
    }
    console.log('✅ Created sample tags');

    // Create sample notes for demo user
    const sampleNotes = [
      {
        title: 'Welcome to NoteHub! 👋',
        body: `# Welcome to NoteHub!

This is your personal note-taking application. Here's what you can do:

## Features
- **Create notes** with rich markdown support
- **Organize** with tags and favorites
- **Share** notes with other users
- **Search** across all your notes
- **Dark mode** for comfortable viewing

## Markdown Support
You can use *italic*, **bold**, and \`code\` formatting.

\`\`\`javascript
// Code blocks are supported too!
const greeting = "Hello, NoteHub!";
console.log(greeting);
\`\`\`

Get started by creating your first note! 🚀`,
        pinned: true,
        favorite: true,
        tags: ['important', 'ideas'],
      },
      {
        title: 'Meeting Notes Template',
        body: `# Meeting Notes

**Date:** [Date]
**Attendees:** [Names]
**Topic:** [Meeting Topic]

## Agenda
1. Item 1
2. Item 2
3. Item 3

## Discussion Points
- Point 1
- Point 2

## Action Items
- [ ] Action 1 - @person - Due date
- [ ] Action 2 - @person - Due date

## Next Steps
[Summary of next steps]`,
        tags: ['work', 'todo'],
      },
      {
        title: 'Quick Ideas',
        body: `# 💡 Ideas

- Build a personal website
- Learn a new programming language
- Start a side project
- Read more books
- Travel to a new place`,
        favorite: true,
        tags: ['personal', 'ideas'],
      },
    ];

    for (const noteData of sampleNotes) {
      const existing = await db.queryOne(`SELECT id FROM notes WHERE title = ? AND owner_id = ?`, [
        noteData.title,
        demoUserId,
      ]);

      if (!existing) {
        const result = await db.run(
          `INSERT INTO notes (title, body, pinned, favorite, owner_id) VALUES (?, ?, ?, ?, ?)`,
          [
            noteData.title,
            noteData.body,
            noteData.pinned ? 1 : 0,
            noteData.favorite ? 1 : 0,
            demoUserId,
          ],
        );

        // Add tags
        if (noteData.tags) {
          for (const tagName of noteData.tags) {
            const tag = await db.queryOne(`SELECT id FROM tags WHERE name = ?`, [tagName]);
            if (tag) {
              await db.run(`INSERT OR IGNORE INTO note_tag (note_id, tag_id) VALUES (?, ?)`, [
                result.insertId,
                tag.id,
              ]);
            }
          }
        }
      }
    }
    console.log('✅ Created sample notes');

    // Create sample tasks for demo user
    const sampleTasks = [
      {
        title: 'Review project requirements',
        description: 'Go through the project documentation and note down key requirements.',
        priority: 'high',
        due_date: new Date(Date.now() + 2 * 24 * 60 * 60 * 1000).toISOString(), // 2 days from now
      },
      {
        title: 'Set up development environment',
        description: 'Install necessary tools and configure the development environment.',
        priority: 'medium',
        completed: true,
      },
      {
        title: 'Write documentation',
        description: 'Create comprehensive documentation for the project.',
        priority: 'low',
        due_date: new Date(Date.now() + 7 * 24 * 60 * 60 * 1000).toISOString(), // 7 days from now
      },
    ];

    for (const taskData of sampleTasks) {
      const existing = await db.queryOne(`SELECT id FROM tasks WHERE title = ? AND owner_id = ?`, [
        taskData.title,
        demoUserId,
      ]);

      if (!existing) {
        await db.run(
          `INSERT INTO tasks (title, description, priority, completed, due_date, owner_id) VALUES (?, ?, ?, ?, ?, ?)`,
          [
            taskData.title,
            taskData.description,
            taskData.priority,
            taskData.completed ? 1 : 0,
            taskData.due_date || null,
            demoUserId,
          ],
        );
      }
    }
    console.log('✅ Created sample tasks');

    console.log('\n🎉 Database seeded successfully!\n');
    console.log('Default credentials:');
    console.log('  Admin: admin / [NOTES_ADMIN_PASSWORD from .env]');
    console.log('  Demo:  demo / Demo12345678!\n');

    await db.close();
    process.exit(0);
  } catch (error) {
    console.error('❌ Seed error:', error);
    process.exit(1);
  }
}

seed();
